File: cpp/SolverAiComputeInput.cpp
#include "SolverAiComputeInput.h"

SolverAiComputeInput::SolverAiComputeInput(int problem_id) : problem_id(problem_id) {}

void SolverAiComputeInput::addInput(const std::string &name, float Min, float Max, bool isConstant, bool isInteger)
{
    inputs[name] = {{"Min", Min}, {"Max", Max}, {"Constant", isConstant}, {"Integer", isInteger}};
}

void SolverAiComputeInput::addConstraint(const std::string &name, CONSTRAINT operation, float Value1, float Value2)
{
    std::string op;
    switch (operation)
    {
    case SMALLER_THAN:
        op = "smaller than";
        break;
    case GREATER_THAN:
        op = "greater than";
        break;
    case EQUAL_TO:
        op = "equal to";
        break;
    case INSIDE_RANGE:
        op = "inside range";
        break;
    case OUTSIDE_RANGE:
        op = "outside range";
        break;
    }
    constraints[name] = {{"Operation", op}, {"Value1", Value1}, {"Value2", Value2}};
}

void SolverAiComputeInput::addObjective(const std::string &name, OBJECTIVE operation)
{
    std::string op = operation == MINIMIZE ? "minimize" : "maximize";
    objectives[name] = {{"Operation", op}};
}

nlohmann::json SolverAiComputeInput::getJson() const
{
    return {
        {"id", problem_id},
        {"inputs", inputs},
        {"constraints", constraints},
        {"objectives", objectives}};
}

File: python/SolverAiComputeInput.py
from enum import Enum
from json import dumps


class CONSTRAINT(Enum):
    SMALLER_THAN = 'smaller than'
    GREATER_THAN = 'greater than'
    EQUAL_TO = 'equal to'
    INSIDE_RANGE = 'inside range'
    OUTSIDE_RANGE = 'outside range'


class OBJECTIVE(Enum):
    MINIMIZE = 'minimize'
    MAXIMIZE = 'maximize'


class SolverAiComputeInput:

    def __init__(self, problem_id):
        self.problem_id = problem_id
        self.inputs = {}
        self.constraints = {}
        self.objectives = {}

        self.setSolverSetup()

    def setSolverSetup(
        self,
        includeLeastInfeasible: bool = False,
        solutionQuality: int = 1
    ):
        self.solverSetup = {
            'includeLeastInfeasible': int(includeLeastInfeasible),
            'solutionQuality': solutionQuality
        }

    def addInput(self,
                 name, Min, Max=0,
                 is_constant=False, is_integer=False):
        if Min == Max:
            is_constant = True
        self.inputs[name] = {
            'Min': Min,
            'Max': Max,
            'Constant': is_constant,
            'Integer': is_integer
        }

    def addConstraint(self, name, operation, value1, value2=0):
        self.constraints[name] = {
            'Operation': operation.value,
            'Value1': value1,
            'Value2': value2
        }

    def addObjective(self, name, operation):
        self.objectives[name] = {
            'Operation': operation.value}

    def getJson(self):
        return dumps({
            'id': self.problem_id,
            'solverSetup': self.solverSetup,
            'inputs': self.inputs,
            'constraints': self.constraints,
            'objectives': self.objectives
        })
